### ISSUE TITLE
Nullpointer @ ThreeDS2Service.INSTANCE.initialize

### ISSUE THREAD

**Issue by sasindroid** [Adyen/adyen-3ds2-android#10]
Hello, when I try to initialize the ThreeDS2Service on my demo android app I get a NullPointerException. I use some dummy config params to initialize. Any thoughts?

Here is my code:
`
    String directoryServerId = "A000000003";
    String directoryServerPublicKey = "eyJrdHkiOiJSU0EiLCJlIjoiQVFBQiIsIm4iOiI4VFBxZkFQ==...";

    ConfigParameters configParameters = new AdyenConfigParameters.Builder(
        directoryServerId, directoryServerPublicKey
    ).build();

    Transaction transaction = null;

    try {
      ThreeDS2Service.INSTANCE.initialize(this, configParameters, null, null);
      // ... rest of the code
    } catch (Exception e) {
      e.printStackTrace();
    }
`

Stacktrace:

`
java.lang.NullPointerException
at java.util.Objects.requireNonNull(Objects.java:203)
at java.util.Arrays$ArrayList.<init>(Arrays.java:3741)
at java.util.Arrays.asList(Arrays.java:3728)
at com.adyen.threeds2.internal.c.a.d.c(Unknown Source:71)
at com.adyen.threeds2.internal.c.a.d.a(Unknown Source:37)
at com.adyen.threeds2.internal.c.b.a(Unknown Source:101)
at com.adyen.threeds2.internal.c.b.<init>(Unknown Source:60)
at com.adyen.threeds2.internal.c.b.a(Unknown Source:51)
at com.adyen.threeds2.internal.g.initialize(Unknown Source:93)
at com.sasi.adyen3ds2.MainActivity.initiateIdentityShopper(MainActivity.java:39)
at java.lang.reflect.Method.invoke(Native Method)
at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:385)
at android.view.View.performClick(View.java:7333)
at android.widget.TextView.performClick(TextView.java:14160)
at android.view.View.performClickInternal(View.java:7299)
at android.view.View.access$3200(View.java:846)
at android.view.View$PerformClick.run(View.java:27773)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:6990)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445)
`

**Comment by timon-langlotz** [comment]
Hi @sasindroid,

where are you retrieving the `directoryServerId` and `directoryServerPublicKey` from?

Cheers,
Timon

**Comment by sasindroid** [comment]
@timon-langlotz - I just picked the values from this link - https://docs.adyen.com/classic-integration/3d-secure/native-3ds2/android-sdk-integration

I am not expecting a successful response, because I know the values aren't real (dummy). But, I don't expect a null pointer and will be good to return a valid exception.

What do you think?

**Comment by timon-langlotz** [comment]
Thanks for the details!

I agree that there shouldn't be a NPE, but rather a documented Exception.
We'll look into it!

**Comment by sasindroid** [comment]
Hi @timon-langlotz  - Thank you. Please keep me updated.

In the mean time, we are trying to see if this SDK fits our requirements for 3DS2 integration.
Is there any work around for me to move forward using this SDK? Do you think I should be ok if I use a valid `directoryServerId/directoryServerPublicKey`?

**Comment by timon-langlotz** [comment]
With a valid `directoryServerId` and `directoryServerPublicKey`, you should be able to move forward.

Do you already have an Adyen test account? If not, you can request one [here](https://www.adyen.com/signup).

You will then need to use our [API](https://docs.adyen.com/checkout/3d-secure/native-3ds2/api-integration) to retrieve the initialisation data for the SDK. There's generally two approaches:
- Integration with Checkout API (recommended)
- Integration with Classic API

Hope this helps!
Cheers,
Timon

**Comment by timon-langlotz** [comment]
Hi @sasindroid,

With:
```
String directoryServerId = "A000000003";
String directoryServerPublicKey = "eyJrdHkiOiJSU0EiLCJlIjoiQVFBQiIsIm4iOiI4VFBxZkFQ==...";
```
I could identify an issue related to base64, but did not encounter a NPE.

Are you sure you're encountering the NPE during `ThreeDS2Service.initialize()`?
Or does it happen during `ThreeDS2Service.createTransaction()` or `Transaction.doChallenge()`?

Cheers,
Timon

**Comment by sasindroid** [comment]
Hi @timon-langlotz - Sorry I missed the your last post.

I tried providing a valid `directoryServerId` and `directoryServerPublicKey` and still get a NPE.

Here is my playground repo: [https://github.com/sasindroid/Adyen3DS2Demo](url)

```
java.lang.NullPointerException
at java.util.Objects.requireNonNull(Objects.java:203)
at java.util.Arrays$ArrayList.<init>(Arrays.java:3741)
at java.util.Arrays.asList(Arrays.java:3728)
at com.adyen.threeds2.internal.c.a.d.c(Unknown Source:71)
at com.adyen.threeds2.internal.c.a.d.a(Unknown Source:37)
at com.adyen.threeds2.internal.c.b.a(Unknown Source:101)
at com.adyen.threeds2.internal.c.b.<init>(Unknown Source:60)
at com.adyen.threeds2.internal.c.b.a(Unknown Source:51)
at com.adyen.threeds2.internal.g.initialize(Unknown Source:93)
at com.sasi.adyen3ds2.MainActivity.initiateIdentityShopper(MainActivity.java:35)
```

**Comment by timon-langlotz** [comment]
Hi @sasindroid,

how did you obtain the directory server ID and public key?

Cheers,
Timon

**Comment by sasindroid** [comment]
Hi @timon-langlotz - As per documentation, I get the server Id & public key from the /authorise response.

This is the curl command as per documentation

```
    curl https://pal-test.adyen.com/pal/servlet/Payment/v46/authorise \
    -H "X-API-key: <fill in your api key>" \
    -H "Content-Type: application/json" \
    -d '{
        "amount":{  
        "currency":"EUR",
        "value":1500
        },
        "merchantAccount":"<fill in your merchant account>",
        "reference":"TEST",
        "threeDS2RequestData":{  
          "deviceChannel":"app"
        },
        "additionalData": {
            "card.encrypted.json": "<fill in your encrypted card json>"
        }
      }'  
```

**Comment by sasindroid** [comment]
Hi @timon-langlotz - I also tried the values from api explorer here - [https://docs.adyen.com/api-explorer/#/Payment/v46/authorise__example_authorise-3d-secure-2](url) and yet the same result.

**Comment by timon-langlotz** [comment]
I tried as well with the API explorer values, and it works for me. I used:
- directoryServerId = "F013371337"
- directoryServerPublicKey = "eyJrdHkiOiJSU0EiLCJlIjoiQVFBQiIsIm4iOiI4VFBxZkFOWk4xSUEzcHFuMkdhUVZjZ1g4LUpWZ1Y0M2diWURtYmdTY0N5SkVSN3lPWEJqQmQyaTBEcVFBQWpVUVBXVUxZU1FsRFRKYm91bVB1aXVoeVMxUHN2NTM4UHBRRnEySkNaSERkaV85WThVZG9hbmlrU095c2NHQWtBVmJJWHA5cnVOSm1wTTBwZ0s5VGxJSWVHYlE3ZEJaR01OQVJLQXRKeTY3dVlvbVpXV0ZBbWpwM2d4SDVzNzdCR2xkaE9RUVlQTFdybDdyS0pLQlUwNm1tZlktUDNpazk5MmtPUTNEak02bHR2WmNvLThET2RCR0RKYmdWRGFmb29LUnVNd2NUTXhDdTRWYWpyNmQyZkppVXlqNUYzcVBrYng4WDl6a1c3UmlxVno2SU1qdE54NzZicmg3aU9Vd2JiWmoxYWF6VG1GQ2xEb0dyY2JxOV80NncifQ=="

Could you try with those?

**Comment by sasindroid** [comment]
@timon-langlotz - Yes, I did try with the same values before and now. I still get a NPE.

Can you pls have a look at my [demo repo](https://github.com/sasindroid/Adyen3DS2Demo) and suggest the changes?

**Comment by timon-langlotz** [comment]
Hi @sasindroid,

apologies, we discovered a bug in our SDK regarding checking of permissions.
We try to get all permissions as a list, and this line fails, since `packageInfo.requestedPermissions` is `null`:
`Arrays.asList(packageInfo.requestedPermissions)`
As a workaround, you can add any permission to your sample project, e.g.:
`<uses-permission android:name="android.permission.INTERNET" />`

We'll make a new release in the next days.

Regards,
Timon

**Comment by sasindroid** [comment]
Hi @timon-langlotz - Alright, it works now. Thanks.

Should I leave this issue opened until the new release?

**Comment by timon-langlotz** [comment]
Hi @sasindroid,

this issue should be fixed with our latest release (2.1.0-rc04).

Cheers!
